scene/animation/animation_node_state_machine.cpp

```diff
@@ -792,11 +799,23 @@ Vector2 AnimationNodeStateMachine::get_graph_offset() const {
 	return graph_offset;
 }
 
+void AnimationNodeStateMachine::_on_playback_state_changed(const StringName &p_from_state, const StringName &p_to_state) {
+
+	Ref<AnimationNodeStateMachinePlayback> playback = get_parameter(this->playback);
+	ERR_FAIL_COND(playback.is_null());
+
+	emit_signal("state_changed", p_from_state, p_to_state);
+}
+
 float AnimationNodeStateMachine::process(float p_time, bool p_seek) {
 
 	Ref<AnimationNodeStateMachinePlayback> playback = get_parameter(this->playback);
 	ERR_FAIL_COND_V(playback.is_null(), 0.0);
 
+	if (!playback->is_connected("state_changed", this, "_on_playback_state_changed")) {
```

Comment:
I wonder whether it's possible to simply connect the `state_changed` signal whenever `AnimationNodeStateMachinePlayback` is assigned to state machine rather than checking this in process, though I don't see an easy way to assign it here. Perhaps it should be handled somewhere in `_set('parameters/playback', playback)`.

`AnimationNodeStateMachinePlayback` is also not shared between instances by default, so I wonder whether it's by design and if it makes sense to create and assign a different playback resource at all:
```cpp
AnimationNodeStateMachinePlayback::AnimationNodeStateMachinePlayback() {
    set_local_to_scene(true); //only one per instanced scene
    // ...
}

Reply:
This seems to be the only place that instances playback resource:
```cpp
Variant AnimationNodeStateMachine::get_parameter_default_value(const StringName &p_parameter) const {

	if (p_parameter == playback) {
		Ref<AnimationNodeStateMachinePlayback> p;
		p.instance();
                // Connect here?
		// p->connect("state_changed", this, "_on_playback_state_changed"));
		return p;
	} else {
		return false; //advance condition
	}
}
```

Reply:
Ok I'll try a version where it connects here and run my tests to see

Reply:
@Xrayez @jedStevens I've got a working version where the connection is made in `AnimationNodeStateMachine::get_parameter_default_value()` and also in `AnimationNodeStateMachine::_set()`. Note that this requires `AnimationNodeStateMachine::get_parameter_default_value()` to no longer be a `const` method.

Is it possible for me to push my changes to this PR or do I need to open a new one?